src/backend/replication/walsender.c

```diff
@@ -258,7 +259,7 @@ static void WalSndPrepareWrite(LogicalDecodingContext *ctx, XLogRecPtr lsn, Tran
 static void WalSndWriteData(LogicalDecodingContext *ctx, XLogRecPtr lsn, TransactionId xid, bool last_write);
 static void WalSndUpdateProgress(LogicalDecodingContext *ctx, XLogRecPtr lsn, TransactionId xid,
 								 bool skipped_xact);
-static XLogRecPtr WalSndWaitForWal(XLogRecPtr loc);
+XLogRecPtr WalSndWaitForWal(XLogRecPtr loc);
```

Comment:
I'm not sure if it's worth exporting this function to use in neon extension.

I looked at `NeonOnDemandXLogReaderRoutines` code, it accepts routine with:
```
XL_ROUTINE(.page_read = logical_read_xlog_page,
		.segment_open = WalSndSegmentOpen,
		.segment_close = wal_segment_close),
```

I have an idea that we can take `logical_read_xlog_page` from it as a `.page_read` and use it whenever we want to read recent WAL.
Otherwise, if we're reading non-recent WAL from safekeepers, we don't need to wait for WAL at all because it's not present locally at all and we need to fetch it from the safekeepers.
WDYT?

Reply:
If I understand correctly, you want to rely on the fact that `XLogReader` is already filled in with `logical_read_xlog_page`, and take the pointer to that and call it whenever we have local WAL? I think it could work but is also more complicated. I think in the case you describe, `WalSndWaitForWal` will immediately return after this if statement:
```
	if (RecentFlushPtr != InvalidXLogRecPtr &&
		loc <= RecentFlushPtr)
		return RecentFlushPtr;
```
So it's no big deal if we call this function when we need to go to the safekeeper.

Reply:
I'm thinking about reducing diff in our postgres and possibly compatibility with the future releases. Also I'm a bit suspicious that `WalSndWaitForWal` does a lot of stuff, such as juggling postgres timelines (we don't support them), was thinking that it would be nice to leave it fully to walsender to figure out everything without calling `WalSndWaitForWal` directly.

But it's probably quite complicated, current code should work, we can come back to this later.